### PR TITLE
Add ingress configuration support

### LINKED ISSUE #64: Serving Operator should provide an option that all ingress gateways are configured
If a user creates a custom ingress gateway, istio may select a gateway at random (https://github.com/knative/serving/issues/5163#issuecomment-521815915). The operator should provide a way to ensure that the ingress gateway always works.

This relies a bit on https://github.com/knative/serving/issues/5172.

I believe this could work by montoring gateways and adjusting the config-istio to account for the new gateway (https://github.com/knative/serving/blob/b7dfe3fdb80ffac7f7a05c20787400ce51aad2b1/config/config-istio.yaml#L45)

This issue is a clone of an existing issue in knative/serving-operator: https://github.com/knative/serving-operator/issues/156

### PR DESCRIPTION
**Pull request by houshengbo**
Co-authored-by: Markus Thömmes \<mthoemme@redhat.com\>

Fixes #64 

## Proposed Changes

* Added Kourier, Contour and Istio as the available ingresses.
* Read the ingress manifests under kodata/ingress/\<version\>.
* Serving component aligns the ingress at the same minor version.

**Release Note**

```release-note

```

### CODE REVIEW

**Review by nak3 — COMMENTED** [review]
There are still many of Istio specific code like:
https://github.com/knative/operator/blob/18aca414e5109ee3f66b4431430616c65ef799ab/pkg/reconciler/knativeserving/common/ha.go#L38
https://github.com/knative/operator/blob/18aca414e5109ee3f66b4431430616c65ef799ab/pkg/reconciler/knativeserving/common/gateway.go#L26
It would be a problem if we use non-istio Ingress, wouldn't it?

**Comment by markusthoemmes** [comment]
Has the feature doc been updated to reflect the necessary changes and have those been discussed and accepted?

File: pkg/reconciler/common/releases.go
Comment (jcrossley3) [inline pkg/reconciler/common/releases.go:228]: It's surprising to see any mention of "ingress" in the common package since it's strictly related to `KnativeServing`. I wouldn't expect any function operating on a `KComponent` to know anything about an ingress.

Comment (houshengbo) [reply pkg/reconciler/common/releases.go]: I moved the functions into ingress package. They are called by the knativeserving reconciler.

File: pkg/apis/operator/v1alpha1/knativeserving_types.go
Comment (jcrossley3) [inline pkg/apis/operator/v1alpha1/knativeserving_types.go:122]:
```suggestion
	Contour ContourIngressConfiguration `json:"contour"`
```

File: pkg/reconciler/knativeserving/ingress/ingress.go
Comment (nak3) [inline pkg/reconciler/knativeserving/ingress/ingress.go:37]: I might be wrong, but is this `return true` correct? If the resources have no `providerLabel`, operator manages it? :thinking:

**Review by nak3 — COMMENTED** [review]
`config/300-serving.yaml` also needs to be updated?

File: pkg/reconciler/knativeserving/ingress/ingress.go
Comment (houshengbo) [reply pkg/reconciler/knativeserving/ingress/ingress.go:37]: If the providerLabel is NOT found, the resource is then from the knative serving, so we need to keep it by returning true. :-)

**Comment by evankanderson** [comment]
Having just checked in #360 , I wonder whether we'd prefer to simply ship different config directories for serving + each ingress plugin.

My main concern having put that together was making sure that if you tried to pin to a particular release, you'd get the same manifests even if you upgraded the operator, by using the logic here:

https://github.com/knative/operator/blob/fd53bd554e19788c081861f1c9206e059a7c053c/pkg/packages/release.go#L170

Sadly, I forgot to write a comment; but the logic for "additional packages" is to use the most-recent release of the additional package which is _older_ than the selected patch version, if one exists, otherwise to use the ".0" patch-release. This does mean that the main package might need to create a new patch release if it wants to expose a new patch release in an additional component, but provides a repeatable way (assuming forward flow of time) to align the manifests.

**Comment by evankanderson** [comment]
@markusthoemmes -- I'd love your thoughts on using "latest minor" for additional packages or "snapshotted files" like I suggested in #360. A sample config if we wanted to use the fetcher to maintain alternate ingress configs would look like this gist:

https://gist.github.com/evankanderson/45fc3d45f487c2de77b0da5b373d5f29

**Comment by evankanderson** [comment]
(also @houshengbo , but I was chatting with him on slack)

**Comment by houshengbo** [comment]
@evankanderson @markusthoemmes 
PLZ check the design doc: https://docs.google.com/document/d/1AQBGCIoUAB8Whg02MVWdujMj8O-OWFvPbccoctMbgJA/edit#
I pasted a snapshot, regarding the structure of the ingress directory in the kodata. That is also what I am implementing in this PR.

About how to align the versions of knative serving and ingress, the answer from me is that we align by the MINOR version.
When user specifies `ma.mi.pa` or `ma.mi` in the spec.version field, operator just picks up the value of `ma.mi`, and locates the `ma.mi` folder under ingress to load all the ingress manifests.

The directory ingress/`ma.mi` saves the LATEST available ingress version available under the MINOR version. We can update it from time to time, when there is new patch release for a certain ingress.

File: pkg/apis/operator/v1alpha1/knativeserving_types.go
Comment (markusthoemmes) [inline pkg/apis/operator/v1alpha1/knativeserving_types.go:156]: Let's drop this as long as we don't ship the manifests.

File: pkg/reconciler/knativeserving/ingress/kourier.go
Comment (markusthoemmes) [inline pkg/reconciler/knativeserving/ingress/kourier.go:30]: No need to make this globally available I think. Private (lowercase) should suffice.

File: pkg/reconciler/knativeserving/ingress/testdata/kodata/ingress/0.18/kourier.yaml
Comment (markusthoemmes) [inline pkg/reconciler/knativeserving/ingress/testdata/kodata/ingress/0.18/kourier.yaml:1]: Should these be symlinks?

File: pkg/reconciler/knativeserving/knativeserving.go
Comment (markusthoemmes) [inline pkg/reconciler/knativeserving/knativeserving.go:120]: Can we come up with a better name? Like `filterDisabledIngresses`?

File: pkg/reconciler/knativeserving/ingress/ingress_test.go
Comment (markusthoemmes) [inline pkg/reconciler/knativeserving/ingress/ingress_test.go:36]: No need for public visibility

File: pkg/reconciler/knativeserving/ingress/ingress_test.go
Comment (markusthoemmes) [inline pkg/reconciler/knativeserving/ingress/ingress_test.go:82]: I don't quite understand what this checks. Can we make the test explicitly check that only istio is loaded by default, Kourier if specified and Istio AND Kourier if both are specified?

**Review by markusthoemmes — COMMENTED** [review]
Looks good overall, but I'd tune down the number of supported Ingresses to just the number we ship, potentially even to just 2 to make the diff quite a bit smaller and more focused.

File: pkg/apis/operator/v1alpha1/knativeserving_types.go
Comment (maximilien) [inline pkg/apis/operator/v1alpha1/knativeserving_types.go:69]: Would `IngressConfigs` be a better name?

File: pkg/reconciler/knativeserving/ingress/gloo.go
Comment (maximilien) [inline pkg/reconciler/knativeserving/ingress/gloo.go:29]: Interesting you return `nil` here and not an empty `[]mf.Transformer{}`...

File: pkg/reconciler/knativeserving/ingress/gloo_test.go
Comment (maximilien) [inline pkg/reconciler/knativeserving/ingress/gloo_test.go:30]: This should fail if nil is returned from code... no?

File: pkg/reconciler/knativeserving/ingress/kong.go
Comment (maximilien) [inline pkg/reconciler/knativeserving/ingress/kong.go:29]: same as for the gloo transformer

File: pkg/reconciler/knativeserving/ingress/testdata/kodata/ingress/0.18/kourier.yaml
Comment (maximilien) [reply pkg/reconciler/knativeserving/ingress/testdata/kodata/ingress/0.18/kourier.yaml:1]: Would love that too 👍

**Review by maximilien — APPROVED** [review]
Left a few comments... still learning this codebase.

File: pkg/reconciler/knativeserving/ingress/gloo.go
Comment (houshengbo) [reply pkg/reconciler/knativeserving/ingress/gloo.go]: If it is nil, then the length of the array is 0.

File: pkg/reconciler/knativeserving/ingress/gloo_test.go
Comment (houshengbo) [reply pkg/reconciler/knativeserving/ingress/gloo_test.go]: This test case is removed.

File: pkg/reconciler/knativeserving/ingress/gloo.go
Comment (markusthoemmes) [reply pkg/reconciler/knativeserving/ingress/gloo.go]: Idiomatic Golang allows for transparent use of `nil` slices. Their len is 0 and you can `range` over them no problem. Allows for saving the allocations that returning an empty slice forces.

File: pkg/reconciler/knativeserving/ingress/ingress_test.go
Comment (houshengbo) [reply pkg/reconciler/knativeserving/ingress/ingress_test.go:82]: This test cases check whether it can get exactly the same number of resources available in the ingress directory. I change the name into `expectedResourcesNum`.
To verify if several specified ingresses are loaded, I will add them into the test case `TestTransformers`.

**Comment by nak3** [comment]
I tested with Kourier but it still does not work if I deployed operator in non-`knative-serving` namespace.
The reason is that Kourier gateway find the controller by the `address` in kourier-bootstrap CM:

https://github.com/knative/operator/pull/312/files#diff-0aac84f00c17a91f67fcac679f691c8ac69fadf9379b44070a6b6fa4843cfaeaR352
```
address: "kourier-control.knative-serving"    # The `knative-serving` is a namespace of kourier-control.
````

So the address needs to be updated somehow.

**Comment by nak3** [comment]
(With said, I think this needs to be fixed in Kourier. It should be configurable by env value or something easier.)

File: pkg/reconciler/knativeserving/ingress/kourier.go
Comment (markusthoemmes) [inline pkg/reconciler/knativeserving/ingress/kourier.go:44]: This filter seems odd. Shouldn't this match a very specific deployment instead?

**Review by markusthoemmes — COMMENTED** [review]
Agree with @nak3. We should invest in writing a transform that mucks around with the config potentially. We can ship that after this initially lands.

This is why I wanted to just do **one** ingress at first. To not block things too much on specialties of the other ingresses and to just land the concept.

**Comment by houshengbo** [comment]
@markusthoemmes @nak3 It is possible to write a transformer working on the ConfigMap for Kourier, but would the CM like `kourier-bootstrap` have a fixed structure for data section? It looks like we need dig into multiple layers down to locate `envoy-bootstrap.yaml.static_resources.clusters[1].load_assignment.......address`.
 Operator supports the configuration directly on ConfigMap, with `spec.config` in the CR. If there is a transformer for the ConfigMap, it may lead to some conflict, when we leverage `spec.config` to overwrite the ConfigMap.

**Comment by houshengbo** [comment]
I opened another issue to track Kourier ns: https://github.com/knative/operator/issues/454

**Comment by nak3** [comment]
@houshengbo @markusthoemmes  Thank you! 

One more question, do you think that we don't need to add [contour.yaml](https://github.com/knative-sandbox/net-contour/tree/master/config/contour) for Contour?
At first, I thought that users should install Contour from net-contour or projectcontour as they like (it is same manner with Istio), but Contour's [this job](https://github.com/knative-sandbox/net-contour/blob/release-0.20/config/contour/external.yaml#L2054) is net-contour specific. So users have to deploy the `contour.yaml` from net-contour repo even after enabled contour.

**Comment by houshengbo** [comment]
@nak3 My initial thought is that we stay with net-***.yaml ingress plugin only in operation. Users need to install the ingress by themselves, like istio.
However, Kourier is an exception. They do not need to install Kourier separately, since kourier.yaml has everything in operator, and it is the only artifact we ship. 
It seems now the job you mentioned is a bit awkward. Why is it in the contour.yaml, not net-contour.yaml?
@markusthoemmes Yours?

**Comment by nak3** [comment]
Sorry the job is not net-contour specific. It was misunderstanding.

But it is still so difficult to deploy without using "net-contour's" Contour. net-contour has [`release.yaml`](https://github.com/knative-sandbox/net-contour/releases) which includes all of them so I wonder that it is better to use them.
With said, I understood what you wanted to do. Hmm... :thinking:

**Comment by houshengbo** [comment]
@nak3 If we use the release.yaml consolidating everything of contour, users do not need to care about installing contour? Is it correct?
What if they have already installed contour? Would there be conflict?

**Comment by nak3** [comment]
Alright, LGTM then. I am leaving other reviewers.

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
